Use simple plaintext renderer if mime-type is "text/plain"

dennisschagt opened the pull request:
Fixes https://github.com/newsboat/newsboat/issues/1010 and fixes https://github.com/newsboat/newsboat/issues/1022:
- Tested with https://dennisschagt.nl/tmp/1010-feed/feed.atom (based on archive in https://github.com/newsboat/newsboat/issues/1010#issuecomment-646866506. Removed the `<summary>` tags so the filter script from https://github.com/newsboat/newsboat/issues/1010#issuecomment-646871561 is no longer needed).

Fixes https://github.com/newsboat/newsboat/issues/468 (taken from https://github.com/newsboat/newsboat/issues/468#issuecomment-787450424):
- Tested with https://dennisschagt.nl/tmp/newsboat-issue-468.xml

I also successfully tested this plaintext rendering with some Youtube atom feeds, e.g.:
- https://www.youtube.com/feeds/videos.xml?channel_id=UCsXVk37bltHxD1rDPwtNM8Q
- https://www.youtube.com/feeds/videos.xml?channel_id=UCmxePybUpZj8RRuWz6r8uTQ
- https://www.youtube.com/feeds/videos.xml?channel_id=UCsU15yvILBmnHPeAf4SFVaQ

Minoru reviewed (commented):
Closing three issues with one PR, eh? :)

The code looks pretty good to me. I think there could be some tests for plain-text items, to check that the new private method correctly splits the description into lines. A test for `download-full-page` could be nice as well, if it's easy enough to do. I suspect you're probably fed up with testing—you've done lots of it for this PR—but automated tests it how we keep regressions at bay.

Thanks a lot for working on this! \m/

x4e commented:
Hi, I made a similar PR in #1577 without noticing this one. 

In my PR I went with plain text formatting for all mimetypes that were not either `html`, `xhtml`, `text/html`, or `application/xhtml+xml`. Would that be a good approach to take here? 
It doesn't really make much sense to be using HTML renderer for any mimetype that isn't `text/plain`.

Minoru commented:
@x4e I agree; it makes sense to use HTML renderer only for HTML(-like) types, and render everything else as plain text.

dennisschagt commented on src/itemrenderer.cpp:
```diff
@@ -163,11 +179,17 @@ std::pair<std::string, size_t> item_renderer::to_stfl_list(
+		render_plaintext(body, lines);
+	} else {
+		render_html(cfg, body, lines, links, baseurl, false);
+	}
```
Good point, done via rebase in https://github.com/newsboat/newsboat/pull/1500/commits/16ef8ef7710761b0806f81d714d15a6b7ac5d992

dennisschagt commented on src/itemrenderer.cpp:
```diff
@@ -130,6 +130,22 @@ void render_html(
+{
+	std::string::size_type pos = 0;
+	while (pos < source.size()) {
+		const auto end_of_line = source.find_first_of("\r\n", pos);
```
I guess it makes sense but I see a few question marks due to `\r` characters in articles from youtube feeds.

For example, the `Why Bayes rule is nicer with odds` entry from https://www.youtube.com/feeds/videos.xml?channel_id=UCYO_jab_esuFRV4b17AJtAw, which shows multiple question marks in the lower half of the description:
![image](https://user-images.githubusercontent.com/4629607/114260834-741ce080-99d7-11eb-95e0-1287fe0fc190.png)

dennisschagt replied:
Looks like the description includes the `\r` characters as entity references (`&#13;`) so that explains why libxml2 does not normalize them:
![image](https://user-images.githubusercontent.com/4629607/114263274-d6301280-99e4-11eb-8947-4ef4072f8f6d.png)

Minoru replied:
Got it; thanks for the investigation! Let's leave this as-is, then.

Minoru replied:
By "as-is" I mean "roll back to using `\r\n` (as it was when this discussion was started)". Sorry for the confusion, it's just now that I'm re-reviewing the updated PR that I notice you've changed it to `\n` already.

dennisschagt commented:
> In my PR I went with plain text formatting for all mimetypes that were not either html, xhtml, text/html, or application/xhtml+xml. Would that be a good approach to take here?
It doesn't really make much sense to be using HTML renderer for any mimetype that isn't text/plain.

Sounds good, applied in https://github.com/newsboat/newsboat/pull/1500/commits/7c6894b5a6df1a5449d2a8db9a7d9058c0111734.

dennisschagt commented:
> I think there could be some tests for plain-text items, to check that the new private method correctly splits the description into lines.

Done: https://github.com/newsboat/newsboat/pull/1500/commits/a83f00da9d1ef59775be0b6a5f68ea6d7f8d4649

> A test for download-full-page could be nice as well, if it's easy enough to do.

I think that might take quite some time (would need to start an HTTP server or fake something which can transmit HTTP headers):
https://github.com/newsboat/newsboat/blob/77cd6b96609688ba8910799623f50c70053e9baa/src/rssparser.cpp#L515-L538

x4e commented on src/itemrenderer.cpp:
```diff
@@ -9,6 +10,17 @@
 
+bool should_render_as_html(const std::string mime_type)
+{
+	std::set<std::string> html_mime_types = {
```
Would `static` be appropriate here to prevent the set being constructed on each call?

dennisschagt replied:
Good idea, amended into https://github.com/newsboat/newsboat/pull/1500/commits/81e72c3071bbb0fa14d288f87cebf988413f7c01

x4e commented on src/itemrenderer.cpp:
```diff
@@ -130,17 +142,40 @@ void render_html(
+		const auto end_of_line = source.find_first_of("\n", pos);
+		const std::string line = source.substr(pos, end_of_line - pos);
+		lines.push_back(std::make_pair(LineType::wrappable, line));
+		if (end_of_line == std::string::npos) {
```
If `end_of_line` is `npos` won't this break the `substr` call above? Should this check not be above the substr call?

Minoru replied:
It won't break `substr` because `substr` only returns as much data as the source string has.

Minoru reviewed (approved):
I think this is ready to be merged after you've rolled back the `\r\n`→`\n` change (see https://github.com/newsboat/newsboat/pull/1500#discussion_r611063407; sorry about the churn!)

Thanks for fixing those three bugs! :rocket:

dennisschagt commented:
Thank you both for the reviews!

> I think this is ready to be merged after you've rolled back the \r\n→\n change

I've dropped the commit and added a new commit to handle `\r\n` line-endings: https://github.com/newsboat/newsboat/pull/1500/commits/094180e51c68e429d03da6369136f44520b86c51

Minoru reviewed (approved):
Thanks! :+1:
